src/vn/edu/rmit/kuri/output/Display.java

```diff
@@ -24,6 +34,147 @@ public static void tabular(TestSummary summary, Metric metric, ResultType result
     //|         .....       |     .....     |
     //+---------------------+---------------+
     System.out.println("To be implemented");
+
+    // Extract needed values based on the metric
+    // Stored data in a 2D ArrayList
+    // Data from the same group will be in the same ArrayList
+    ArrayList<ArrayList<Integer>> value = new ArrayList<>();
+    for (int i = 0; i < summary.size(); i++) {
+      ArrayList<Integer> valueGroup = new ArrayList<>();
+      for (int j = 0; j <summary.get(i).size(); j++)
+      {
+        switch (metric) {
+          case CASES -> {
+            int k = summary.get(i).get(j).getNewCases();
+            valueGroup.add(k);
+          }
+          case DEATHS -> {
+            int k = summary.get(i).get(j).getNewDeaths();
+            valueGroup.add(k);
+          }
+          case VACCINATIONS -> {
+            int k = summary.get(i).get(j).getNewVaccinations();
+            valueGroup.add(k);
+          }
+        }
+      }
+      value.add(valueGroup);
+    }
+
+    // Calculate the value to display based on the resultType
+    // Each group's value will be stored as an integer in an ArrayList
+    ArrayList<Integer> valueForDisplay = new ArrayList<>();
+    int valueOfEachGroup = 0;
+    for (int i = 0; i < value.size(); i++) {
+      for (int j = 0; j < value.get(i).size(); j++) {
+        valueOfEachGroup += value.get(i).get(j);
+        switch (resultType) {
+          case CUMULATIVE -> {
+            if (j == value.get(i).size() - 1) {
+              // do not reset value for cumulative results
+              valueForDisplay.add(valueOfEachGroup);
+            }
+          }
+          case NEW_PER_PERIOD -> {
+            if (j == value.get(i).size() - 1) {
+              valueForDisplay.add(valueOfEachGroup);
+              // reset value for the next group when reach the last value
+              valueOfEachGroup = 0;
+            }
+          }
+        }
+      }
+    }
+
+    // Store each group's range in an ArrayList
+    ArrayList<String> range = new ArrayList<>();
+    for (int i = 0; i < summary.size(); i++) {
+      int groupSize = summary.get(i).size();
+      String startDate = summary.get(i).get(0).getDate().toString();
+      String endDate = summary.get(i).get(groupSize - 1).getDate().toString();
+      if (groupSize > 1) {
+        range.add(startDate + " - " + endDate);
+      } else {
+        range.add(startDate);
+      }
+    }
+
+    // Methods for creating table
```

Comment:
Unclear comment. It should be: "Nested interface containing methods for creating a terminal-based table"